Create SPECIFICATION.md (#5069)
awsdocs/aws-doc-sdk-examples: applications/feedback_sentiment_analyzer/SPECIFICATION.md

> # FSA Tech spec

Reviewer: ```suggestion
# Feedback Sentiment Analyzer (FSA) technical specification
```

Author: With PAM, we used `tech spec`. I would prefer `technical specification` but didn't want to break the established way.

> This document contains the technical specifications for Feedback Sentiment Analyzer (FSA), a sample application that showcases AWS services and SDKs.

Reviewer: Can we specify the services and SDKs here?

Author: Link back to README.md.

> It explains:

Reviewer: Architecture should be included as well with a diagram.

Author: Add before Inputs and Outputs

* Application inputs and outputs
* Underlying AWS components and their configurations
* Key cross-service integration details

### Table of Contents

>   * [Inputs](#inputs)

Reviewer: You might want to review David's feedback on your PAM spec changes. There's similarities.

  * [Outputs](#outputs)
  * [Diagram](#diagram)
  * [Input routing](#input-routing)
  * [Step Function configuration](#step-function-configuration)
    + [State machine Lambda functions](#state-machine-lambda-functions)
      - [Diagram](#diagram-1)
      * [Function inputs and outputs](#function-inputs-and-outputs)
        + [ExtractText](#extracttext)
        + [AnalyzeSentiment](#analyzesentiment)
        + [TranslateText](#translatetext)
        + [SynthesizeAudio](#synthesizeaudio)
  * [GetFeedback Lambda function](#getfeedback-lambda-function-1)
  * [Processing S3 events with EventBridge](#processing-s3-events-with-eventbridge)
  * [Managing items in DynamoDB](#managing-items-in-dynamodb)
  * [Other FSA material](#other-fsa-material)

---

> ## Inputs

Reviewer: These inputs and outputs don't seem very helpful to me as their own section. The lambda specs definitely need input shape and output shape, but the content here might fit better within and end-to-end description of how the app works.

Author: User actions and application behavior.

This application receives three inputs from the frontend.
* authenticate user (uses `Cognito`)
* load page (uses `DynamoDB:GetItems`)
* upload image to S3 (`S3:PutObject`)

## Outputs
This application produces two outputs from the backend:
* put new item in the database (uses `DynamoDB:PutItem`)
* put new synthesized audio to S3 (uses `S3:PutObject`)

## Diagram

> ![img_1.png](img_1.png)

Reviewer: Will this be the arch diagram?

Author: Update w/ more detailed diagram.

---

## Input routing
The application inputs diagrammed above are routed through API Gateway.

| action       | route                    | method | request target                                              | response model                                           |
|--------------|--------------------------|--------|-------------------------------------------------------------|----------------------------------------------------------|

> | Load page    | `/feedback`                | GET    | [GetFeedback Lambda function](#GetFeedback-lambda-function) | `[   {     "text": String,     "audioUrl": String   } ]` |

Reviewer: ```suggestion
| Load page    | `/feedback`                | GET    | [GetFeedback Lambda function](#GetFeedback-lambda-function) | ```{
  "feedback": [
    {
      "sentiment": "POSITIVE",
      "text": "J'adore cet hôtel/",
      "audioUrl": "PXL_20230710_182358532.jpg.mp3",
      "imageUrl": "PXL_20230710_182358532.jpg"
    }
  ]
}``` |
```

Author: route, request body, response body, notes

| Upload image | `/audio/{bucket}/{object}` | PUT    | [State Machine via s3:PutObject API]()                      | void                                                     |

---

## Step Function configuration
When an image is uploaded through `Upload image` input route, a Step Function state machine is triggered.

See [Event processing](#event-processing)
This state machine orchestrates a series of AWS Lambda functions.

### State machine Lambda functions

| trigger                      |        function       | action                                   |
|------------------------------|-----------------------|------------------------------------------|

> | S3:CreateObject event.       | [ExtractText](#ExtractText)      | Extracts text from   S3 object.          |

Reviewer: There's an off-by-one error here.

| `ExtractText` complete.      | [AnalyzeSentiment](#AnalyzeSentiment) | Detect positive or   negative sentiment. |
| `AnalyzeSentiment` complete. | [TranslateText](#TranslateText)    | Translate text to   French.              |
| `TranslateText` complete.    | [SynthesizeAudio](#SynthesizeAudio)  | Synthesize   human-like audio from text. |

> #### Diagram

Author: Diagram at top of section

These functions appear by name in the multi-state ]workflow depicted below. 

Also depicted is where the workflow forks at `ContinueIfPositive` based whether sentiment analyzed is `POSITIVE` or `NEGATIVE`. 

When ran successfully, the state machine  adds a new item to a DynamDB table containing data gathered during the workflow.

![FSA - l'hotellerie.png](..%2F..%2F..%2F..%2F..%2FDownloads%2FFSA%20-%20l%27hotellerie.png)

Logging appears in the AWS Console under Step Functions and (for Lambda functions only) CloudWatch.

## Function inputs and outputs
See below for the required inputs and outputs of each Lambda function.

### ExtractText

> 

Reviewer: Each function needs a description of it's behavior. For example:
> ExtractText is the entry point in a series of AWS Step Functions. It's triggered when an image file is uploaded to the Amazon S3 `media-bucket` created when you run `cdk deploy` in [the README](./README.md). The function uses the inputs described below to run [DetectDocumentText](https://docs.aws.amazon.com/textract/latest/dg/API_DetectDocumentText.html) with Amazon Textract. The extracted words are joined together as one string and returned.

Author: Start w/ bullet points that are super concise, then link to individual sections.

At DynamoDB step, link down to Dynamo section

#### **Input**
Function will use data available on the [S3 event object](https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventBridge.html). 

For example:
```json
{
  "bucket": "my-s3-bucket",
  "region": "us-east-1",
  "object": "foo/bar.png"
}
```
#### **Output**
Function will return a string representing the text extracted.

For example:
```json
THIS HOTEL WAS GREAT
```

> ### AnalyzeSentiment

Author: Add explanation like the one proposed above

#### **Input**
Function will use data available on the [Lambda event object](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-concepts.html#gettingstarted-concepts-event). 

For example:
```json
{
    "source_text": "THIS HOTEL WAS GREAT",
    "region": "us-east-1"
 }
```
#### **Output**
Function will return a string representing the text extracted.

For example:
```json
{
   "sentiment": 'POSITIVE',
   "language_code": 'en'
}
```

> ### TranslateText

Author: Add explanation like the one proposed above

#### **Input**
Function will use data available on the [Lambda event object](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-concepts.html#gettingstarted-concepts-event).

For example:
```json
{
    "source_language_code": "en"
    "region": "us-east-1",
    "extracted_text": "THIS HOTEL WAS GREAT"
}
```
#### **Output**
Function will return a string representing the text translated.

For example:
```json
CET HOTEL ÉTAIT RAVISSANT
```

> ### SynthesizeAudio

Author: Add explanation like the one proposed above

#### **Input**
Function will use data available on the [Lambda event object](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-concepts.html#gettingstarted-concepts-event).

For example:
```json
{
    "bucket": "my-s3-bucket",
    "translated_text": "CET HOTEL ÉTAIT RAVISSANT",
    "region": "us-east-1",
    "object": "comment.png"
 }
```
#### **Output**
Function will return a string representing the key of the newly-synthesized audio file.

For example:
```json
my-s3-bucket/audio.mp3
```
---

## GetFeedback Lambda function

> To display data on previously-uploaded images, the frontend invokes a Lambda function which fetches items from a [DynamoDB table](#DynamoDB).

Reviewer: We should note that this function is hardcoded as part of the CDK deployment and does not need to be implemented by specific languages.

Author: Add explanation like the one proposed above

There is no input.

### **Output**
Function will return a JSON object `feedback` containing items. Each item will contain translated text and a link to its corresponding synthesized audio file.

For example:
```json

> {"feedback": [
>   {
>     "text": String,
>     "audioUrl": String
>   }
> ]}

Reviewer: ```suggestion
{
  "feedback": [
    {
      "sentiment": "POSITIVE",
      "text": "J'adore cet hôtel/",
      "audioUrl": "PXL_20230710_182358532.jpg.mp3",
      "imageUrl": "PXL_20230710_182358532.jpg"
    }
  ]
}
```

```
---
## Processing S3 events with EventBridge
This application relies on an [EventBridge rule](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-rules.html), which triggers the [Step Functions state machine](#step-function-configuration) when new images are uploaded to S3 by the frontend.

> Specifically, the trigger is scoped to `ObjectCreated` events emitted by `my-s3-bucket`:

Reviewer: ```suggestion
Specifically, the trigger is scoped to `ObjectCreated` events emitted by the `media-bucket` created during the CDK deployment:
```

```json
{
  "detail-type": ["Object Created"],
  "source": ["aws.s3"],
  "detail": {
    "bucket": {
      "name": ["my-s3-bucket"]
    }
  }
}
```

---

> ## Managing items in DynamoDB

Reviewer: We should have a general summary of how dynamodb is used.

| action       | route                    | method | request target                                              | response model                                           |
|--------------|--------------------------|--------|-------------------------------------------------------------|----------------------------------------------------------|
| Load page    | `/feedback`                | GET    | [GetFeedback Lambda function](#GetFeedback-lambda-function) | `[   {     "text": String,     "audioUrl": String   } ]` |
| Upload image | `/audio/{bucket}/{object}` | PUT    | [State Machine via s3:PutObject API]()                      | void                                                     |

---

## Step Function configuration
When an image is uploaded through `Upload image` input route, a Step Function state machine is triggered.

See [Event processing](#event-processing)
This state machine orchestrates a series of AWS Lambda functions.

### State machine Lambda functions

| trigger                      |        function       | action                                   |
|------------------------------|-----------------------|------------------------------------------|

> | S3:CreateObject event.       | [ExtractText](#ExtractText)      | Extracts text from   S3 object.          |

Author: ```suggestion
| S3:CreateObject event.       | [ExtractText](#ExtractText)      | Extracts text from S3 object.          |
```
